2/answers.txt

```diff
@@ -0,0 +1,231 @@
+
+Q23: What is the difference between concurrent and parallel execution? What does 'thread safety' mean?
+
+Concurrent execution is when multiple processes are being worked upon at the same time, but not executed at the same time, whereas parallel execution is where 2 or more processes are being executed simultaneously.
+An application can be any number of the above, i.e. concurrent but not parallel, parallel but not concurrent, neither and both. Thread safety applies specifically to multi threaded code. 
+
+Thread safety is about ensuring that data structure manipulated by multiple threads affect the data in the intended way, without causing unexpected changes or damage as a result of multiple entry points to the data structure.
+Typically these are done by implementing race conditions, however you can achieve thread safety without race conditions by implementing, Re-Entrancy, thread-local storage, immutable objects, mutual exclusion and atomic operations.
+
+Q24: If a CPU has 4 cores, how many processes can it execute concurrently?
+
+A CPU with 4 cores can run as many processes "Concurrently" as it has memory to store the not currently executing processes.
+
+Q25: Explain the difference between synchronous and asynchronous code. When would you prefer one technique over the other?
+
+Synchronous code has to be sequentially completed before the next set of instructions can be processed, asynchronous code allow an instruction set to be hung and another started and then returned back to the previous and so on.
+
+You would use asynchronous techniques where you can run two tasks simultaneously, usually when they do not share dependencies and are independant, and you would use synchronous techniques, when you need a task to complete before proceeding.
+
+Q26: Can a linux web-server operate acceptably with 1CPU and 512MB RAM? Can a Windows server do the same?
+
+Although Linux can run on very little RAM as little as 8MB, the act of serving as a webserver typically requires as much as 2GB RAM to functional minimally, depending of course of the load.
+The same is true of windows, except that the native cost of running windows is higher than linux and typically requires as much as 512MB of RAM to function at all.
+
+Q27: Is it bad for an operating system to have assigned all its available memory? When might this occur and why can it be beneficial to performance rather than detrimental?
+
+If an operating system assigns all of its avialable memory then there is no memory remaining for additional processes, namely memory required for memory allocation processes like dynamic memory allocation.
+Things like garbage collection of memory will also look to use memory to function and if there is no memory to assign then it would crash. 
+
+I can find no example of any instance where performance would be in any way shape or form increased by having all memory assigned. almost all memory allocation methods seem to require space to function correctly.
```

Comment:
Most operating systems partition a large portion of memory as a cache.
Memory that has nothing assigned to it is wasted memory. For this reason, Linux caches recently accessed files inside system memory. This provides a significant speedup to file IO compared to having unused memory and requesting files from disk.

It's common to see all your memory utilised in a linux VM. When another application requests memory, the cache will be cleared up and free memory becomes available for the application.